server/src/http.rs

```diff
@@ -116,25 +124,36 @@ impl<Q: QueryExecutor + 'static> Service<Q> {
     fn sql(&self) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
         warp::path!("sql")
             .and(warp::post())
-            // TODO(yingwen): content length limit
-            .and(warp::body::json())
+            .and(warp::header("content-type"))
+            .and(warp::body::content_length_limit(MAX_BODY_SIZE).and(warp::body::bytes()))
             .and(self.with_context())
             .and(self.with_instance())
-            .and_then(|req, ctx, instance| async {
-                // TODO(yingwen): Wrap common logic such as metrics, trace and error log
-                let result = handlers::sql::handle_sql(ctx, instance, req)
-                    .await
-                    .map_err(|e| {
-                        // TODO(yingwen): Maybe truncate and print the sql
-                        error!("Http service Failed to handle sql, err:{}", e);
-                        Box::new(e)
-                    })
-                    .context(HandleRequest);
-                match result {
-                    Ok(res) => Ok(reply::json(&res)),
-                    Err(e) => Err(reject::custom(e)),
-                }
-            })
+            .and_then(
+                |content_type: String, bytes: bytes::Bytes, ctx, instance| async move {
+                    let req = if content_type.to_lowercase().contains("json") {
```

Comment:
Shall we define two const content types:
```
const JSON_CONTENT_TYPE: &str = "application/json";
const TEXT_CONTENT_TYPE: &str = "text/plain";
```

And then put it like:
```rust
let req = if context_type == JSON_CONTENT_TYPE {
    ...
} else if context_type == TEXT_CONTENT_TYPE {
    ...
} else {
   InvalidContextType.context({}).fail()?;
};
```

Reply:
Ok. And shall we define the constants and do check also for `text/plain`?

Reply:
I find a better way deserving a try, but I'm not sure whether it will work:
```
warp::path!("sql")
            .and(warp::post())
            .and(warp::header("content-type"))
            .and(warp::body::content_length_limit(MAX_BODY_SIZE)
            .and(warp::body::json().or(warp::body::bytes())))
            .and(self.with_context())
            .and(self.with_instance())
```

Reply:
After some tries, I find this works:
```rust
        let extract_sql = warp::body::json()
            .or(
                warp::body::bytes().map(|raw_sql: common_types::bytes::Bytes| {
                    Request::from(String::from_utf8_lossy(&raw_sql).to_string())
                }),
            )
            .unify();
        warp::path!("sql")
            .and(warp::post())
            // TODO(yingwen): content length limit
            .and(extract_sql)
            .and(self.with_context())
            .and(self.with_instance())
```

Reply:
Yes, but there is a pattern for the or filter.  Take a look at the example metioned above.

Reply:
> After some tries, I find this works:
> ```rust
>         let extract_sql = warp::body::json()
>             .or(
>                 warp::body::bytes().map(|raw_sql: common_types::bytes::Bytes| {
>                     Request::from(String::from_utf8_lossy(&raw_sql).to_string())
>                 }),
>             )
>             .unify();
>         warp::path!("sql")
>             .and(warp::post())
>             // TODO(yingwen): content length limit
>             .and(extract_sql)
>             .and(self.with_context())
>             .and(self.with_instance())
>             .and_then(|req, ctx, instance| async {
> ```

Take a look at this.